### PR TITLE
Add monotonic_rnnt torch op

### PR DESCRIPTION
**Pull request by SimBe195**
Port of [https://github.com/SimBe195/monotonic-rnnt](https://github.com/SimBe195/monotonic-rnnt) to i6_native_ops. This was originally a fork of [https://github.com/HawkAaron/warp-transducer](https://github.com/HawkAaron/warp-transducer) with the main changes being:

- Changed to monotonic (RNA) label topology
- Using a packed input activation tensor and merged-in softmax operation for more memory efficiency
- Allowing the restriction of the alignment-path-grid to a region around a given target alignment
- Some general refactoring, test cases etc.

### CODE REVIEW

File: i6_native_ops/monotonic_rnnt/__init__.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
+import os
+from typing import Optional
+
+import torch
+from pkg_resources import get_distribution
+from torch.utils.cpp_extension import load
+
+current_file_path = os.path.abspath(__file__)
+current_directory = os.path.dirname(current_file_path)
```
Comment (DanEnergetics) [inline i6_native_ops/monotonic_rnnt/__init__.py]: I don't think these two lines are used

File: i6_native_ops/monotonic_rnnt/__init__.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        label_lengths,
+        alignment,
+        max_shift_from_alignment,
+        blank_label,
+    )
+    assert result is not None
+    return result
+
+
+class MonotonicRNNTLoss(torch.nn.Module):
```
Comment (DanEnergetics) [inline i6_native_ops/monotonic_rnnt/__init__.py]: When does one need this as a torch.nn.Module? Would a definition as a function like above not be enough? This seems to just call `apply` method of the `MonotonicRNNTFunction`

File: i6_native_ops/monotonic_rnnt/__init__.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        ctx,
+        acts: torch.Tensor,
+        labels: torch.Tensor,
+        input_lengths: torch.Tensor,
+        label_lengths: torch.Tensor,
+        alignment: Optional[torch.Tensor] = None,
+        max_shift_from_alignment: int = 0,
+        blank_label: int = 0,
+    ) -> torch.Tensor:
+        assert core is not None
```
Comment (DanEnergetics) [inline i6_native_ops/monotonic_rnnt/__init__.py]: Would this ever be the case? Wouldn't `torch.utils.cpp_extension.load` just throw an error instead of returning `None`?

File: i6_native_ops/monotonic_rnnt/__init__.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+        input_lengths: torch.Tensor,
+        label_lengths: torch.Tensor,
+        alignment: Optional[torch.Tensor] = None,
+        max_shift_from_alignment: int = 0,
+    ) -> torch.Tensor:
+        """Computes the RNNT loss between a sequence of activations and a
+        ground truth labeling.
+        Args:
+            acts:                      A packed 2-D Tensor of logits*. The dimensions should be
+                                       (T_1*(S_1+1) + T_2*(S_2+1) + ... + T_B*(S_B+1), V), where B is the minibatch index,
```
Comment (DanEnergetics) [inline i6_native_ops/monotonic_rnnt/__init__.py]: IMO this indentation makes it less readable and it's not standard I think

**Review by DanEnergetics — CHANGES_REQUESTED** [review]
Just a few small things. Otherwise, I ran it and it all works fine according to your tests. It's nice that you included numerical calculations in them.
A note on docstring formats: We should decide on the format. So far we have a merged submodule with reST and one with Google style formatting...
Btw, do we need to include the license from the original repository like @JackTemaki did it for the warp_rnnt module?

File: i6_native_ops/monotonic_rnnt/__init__.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
+import os
```
Comment (curufinwe) [inline i6_native_ops/monotonic_rnnt/__init__.py:3]: Please add an `__all__` field.

File: setup.py [hunk 6]
```diff
@@ -36,7 +38,13 @@ class MissingTorchInstallationError(Exception):
                 f"{WARP_RNNT_DIR}/core_gather.cu",
             ],
             include_dirs=[WARP_RNNT_DIR],
-        )
+        ),
+        CUDAExtension(
+            name="i6_native_ops.monotonic_rnnt.monotonic_rnnt_core",
```
Comment (curufinwe) [inline setup.py:53]: rename to `i6_native_ops.monotonic_rnnt.core`

**Review by curufinwe — CHANGES_REQUESTED** [review]

**Comment by SimBe195** [comment]
Implemented all the suggestions from @DanEnergetics and @curufinwe now. Thanks for the feedback!

**Comment by curufinwe** [comment]
Sorry for the rename request. I forgot that this is the name of the .so file. Here it would be better to have individual names for each OP. I reverted this.

**Review by curufinwe — APPROVED** [review]

**Review by DanEnergetics — APPROVED** [review]
All looks good to me!

**Review by JackTemaki — APPROVED** [review]
I only did a quick check over the python interface, besides the kind of hard to understand input format of "acts" this seems fine.
